File: update.py
# 每日1:00新更新前一日数据
import os, time
import pandas as pd
import requests, json
from datetime import timedelta
def get_yesterday():
    t = time.localtime(time.time())
    d, m = t.tm_mday - 1,  t.tm_mon
    if d == 0:
        if m in [1,2,4,6,8,9,11]:
            d = 31
        elif m == 3:
            if t.tm_year % 4 == 0:
                d = 29
            else:
                d = 28
        else:
            d = 30
        m -=1
    return m, d
def get_today():
    t = time.localtime(time.time())
    d, m = t.tm_mday,  t.tm_mon
    return m, d

def extract_data(clean):
    a = clean
    wordwide_his = a[a['province'].isnull()]
    prov_his = a[(-a['province'].isnull()) & (a['city'].isnull())]
    city_his = a[-a['province'].isnull() & -a['city'].isnull()]
    wordwide_his.to_csv('lastest/worldwide_history.csv',index=False)
    prov_his.to_csv('lastest/province_history.csv',index=False)
    city_his.to_csv('lastest/city_history.csv',index=False)

m, d = get_yesterday()
M, D = get_today()
yesterday = '2020-%02d-%02d' % (m, d)
today = '2020-%02d-%02d' % (M, D)

# 备份数据
os.system('cp lastest/clean.csv archive/clean' + yesterday + '.backup')
clean = pd.read_csv('lastest/clean.csv')
# 提取头条上一日的数据；实时数据，所以最新一日的还没记录完整
# 例如：20号上午1:00 获取 19号 数据最可靠。
#tturl_orig = 'https://i.snssdk.com/forum/home/v1/info/?forum_id=1656784762444839'
print('getting data ...')
tturl_github = 'https://raw.githubusercontent.com/canghailan/Wuhan-2019-nCoV/master/Wuhan-2019-nCoV.csv'
df = pd.read_csv(tturl_github)
df1 = df[df.date<='2020-04-27']
df2 = df[df.date>'2020-04-27']
df2.columns = ['date',  'countryCode', 'country','provinceCode', 'province', 'cityCode','city', 'confirmed', 'suspected', 'cured', 'dead']
tt = pd.concat([df1, df2])
tt.to_csv('lastest/'+today+'.csv',index=False)
columns = ['date','country','province','city','confirmed','suspected', 'cured', 'dead']
tt = tt[columns]
tt = tt[tt.date == yesterday]

# 是否需要更新：
if max(clean.date) < yesterday:
    clean = pd.concat([clean, tt])
    clean.to_csv('lastest/clean.csv',index=False)
    extract_data(clean)
    os.system('Rscript update.R')
    open('lastest/update.log','a').write('Added '+yesterday+ ' data')
else:
    print('You have updated!')

File: update.R
city <- read.csv('lastest/city_history.csv',stringsAsFactors = F,encoding="UTF-8")
province <- read.csv('lastest/province_history.csv',stringsAsFactors = F,encoding="UTF-8") 
worldwide <- read.csv('lastest/worldwide_history.csv',stringsAsFactors = F,encoding="UTF-8") 

city = city[,c('date','province','city','confirmed','cured','dead')]
colnames(city) = c('time','province','city','cum_confirm','cum_heal','cum_dead')
city$time = as.Date(city$time,format='%Y-%m-%d')
province = province[,c('date','province','confirmed','cured','dead','suspected')]
colnames(province) = c('time','province','cum_confirm','cum_heal','cum_dead','suspected')
province$time = as.Date(province$time,format='%Y-%m-%d')
worldwide = worldwide[,c('date','country','confirmed','cured','dead')]
colnames(worldwide) = c('time','country','cum_confirm','cum_heal','cum_dead')
worldwide$time = as.Date(worldwide$time,format='%Y-%m-%d')

f_p <- function(province) {
  province = sub("省", "", province)
  province = sub("自治区", "", province)
  province = sub("市", "", province)
  province = sub("特别行政区", "", province)
  province = sub("维吾尔", "", province)
  province = sub("壮族", "", province)
  province = sub("回族", "", province)
  return(province)
}

f_c <- function(city) {
  city <- as.character(city)
  city <- sub("市$", "", city)
  city <- sub("地区$","", city)
  city <- sub("省$","", city)
  city <- sub("特别行政区$","", city)
  city <- sub("土家族","", city)
  city <- sub("蒙古族","", city)
  city <- sub("哈尼族","", city)
  city <- sub("布依族","", city)
  city <- gsub(".族","", city)
  city <- sub("自治","", city)
  city[city == '神农架林区'] = '神农架'
  city[city == '甘孜州'] = '甘孜'
  city[city == '凉山州'] = '凉山'
  return(city)
}

trans <- function(city,f) {
  res <- f(city)
  return(res)
}

f1 <- function(city) {
  city <- as.character(city)
  city[city == '未公布来源'] = '地区待确认'
  city[city == '所属地待确认'] = '地区待确认'
  city[city == '未公布来源'] = '地区待确认'
  city[city == '武汉来京'] = '外地来京'
  city[city == '锡林郭勒盟'] = '锡林郭勒'
  city[city == '伊犁哈萨克州'] = '伊犁'
  city[city == '延边朝州'] = '延边'
  return(city)
}
f2 <- function(city) {
  city <- as.character(city)
  city <- sub("区$", "", city)
  return(city)
}
city$city = trans(city$city,f1)
nn = city$province %in% c('北京市','上海市') #只处理北京上海
city[nn,]$city = trans(city[nn,]$city,f2)

province$province = trans(province$province,f_p)
city$province = trans(city$province,f_p)
city$city = trans(city$city,f_c)

Dataframe_2_S3 <- function(data){
  S3 <- list(
    data = city,
    province = province,
    global = worldwide,
    time = max(data$time) 
  )
  class(S3) <- 'nCov2019History'
  return(S3)
}
b = Dataframe_2_S3(worldwide)
saveRDS(b,'dxy_historical_data.rds') # actually, it is github source data
